### PR TITLE
Update setAbility to not modify Special Abilities

### PR DESCRIPTION
**Pull request by Spandamn**
#3231

### CODE REVIEW

File: battle-engine.js [hunk 1]
```diff
@@ -1055,7 +1055,10 @@ class BattlePokemon {
 	setAbility(ability, source, effect, noForce) {
 		if (!this.hp) return false;
 		ability = this.battle.getAbility(ability);
-		let oldAbility = this.ability;
+		let oldAbility = this.ability, cantModify = {battlebond:1, powerconstruct:1, rkssystem:1, shieldsdown:1};
```
Comment (Zarel) [inline battle-engine.js]: I would not put both on the same line.

We only allow double declaration in very specific conditions, and this isn't it.

**Comment by Spandamn** [comment]
oh fuck wait a sec

**Comment by Zarel** [comment]
Martyyyyyy ;_;

**Comment by Marty-D** [comment]
This doesn't actually address the problem; setAbility is called *after* most of the relevant effects have already decided they're going to activate.

**Comment by Spandamn** [comment]
What effects are you referring to? Can you give an example?
Methot effects occur after setAbility is called...
https://github.com/Zarel/PokemonShowdown/blob/master/data/moves.js#L4540 if oldAbility is fa

**Comment by Spandamn** [comment]
(I might be wrong; please correct me wherever I am)

**Comment by Marty-D** [comment]
Everything mentioned in [the PR you linked](https://github.com/Zarel/Pokemon-Showdown/pull/3231#issuecomment-279293682) and probably anything else that uses setAbility.

**Comment by Spandamn** [comment]
well then, its prolly better to update the `bannedAbilities` for each and everything that uses setAbility ... like in Kris' PR

**Comment by urkerab** [comment]
If `setAbility` is the wrong place to check for abilities that can't be removed in any way, then

 1. Why does setAbility check for unoverwritable/unsettable abilities?
 2. Why do most callers of setAbility check that it succeeded?

(I'm particularly worried about the coding for Mummy, which currently seems to be able to overwrite everything except Multitype and Stance Change.)

**Comment by Marty-D** [comment]
I have no idea, I was just saying that if it's going to be done this way it should be done properly. Either that or continue the trend of having the move lists everywhere.

**Comment by urkerab** [comment]
In that case, please can you confirm or deny that:

 - Battle Bond, Comatose, Disguise, Flower Gift, Forecast, Multitype, Power Construct, RKS System, Schooling, Shields Down, Stance Change and Zen Mode cannot be overwritten by Mummy or suppressed by Core Enforcer or Gastro Acid
- That list, plus Truant, cannot be overwritten by Entrainment, Role Play, Simple Beam, Skill Swap or Worry Seed
 - That list, plus Illusion, Imposter and Trace, cannot be copied by Entrainment or Trace
 - That list, plus Illusion, Imposter, Trace and Wonder Guard, cannot be copied by Power of Alchemy, Receiver, Role Play or Skill Swap

Written this way it looks a whole lot simpler than those cumbersome lists of Abilities, so it would be really nice if this was correct.

**Comment by MacChaeger** [comment]
Pretty sure Ultimario has confirmed that all of those are immutable and uncopiable except Flower Gift. http://www.smogon.com/forums/threads/pokemon-sun-moon-battle-mechanics-research.3586701/page-51#post-7112730 would be nice if this could get through (after conflicts are resolved).

**Pull request closed without merging**